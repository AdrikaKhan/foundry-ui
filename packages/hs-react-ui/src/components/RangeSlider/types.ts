export type valueProp = {
  value: number;
  label?: string | number | Node;
  color?: string;
};

export type containerProps = {
  showDomainLabels?: boolean;
  hasHandleLabels?: boolean;
  disabled: boolean;
  beingDragged: boolean;
};

export type handleProps = {
  beingDragged?: boolean;
  color: string;
};

export type handleLabelProps = { velocity?: number };

export type selectedRangeProps = { min: number; max: number; selectedRange: number[] };

export type domainLabelProps = { position: 'left' | 'right' };

export type RangeSliderProps = {
  StyledContainer?: any;
  StyledDragHandle?: any;
  StyledHandleLabel?: any;
  StyledSlideRail?: any;
  StyledSelectedRangeRail?: any;
  StyledDomainLabel?: any;

  showDomainLabels?: boolean;
  showSelectedRange?: boolean;

  motionBlur?: boolean;
  springOnRelease?: boolean;
  debounceInterval?: number;
  axisLock?: 'x' | 'y' | '';
  onDrag?: (val) => void;
  disabled?: boolean;
  min: number;
  max: number;
  values?: number[] | valueProp[];
  testId: string;
};
