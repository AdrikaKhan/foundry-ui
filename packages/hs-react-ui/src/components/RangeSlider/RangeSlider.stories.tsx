import React, { useState, useEffect, forwardRef } from 'react';
import styled from 'styled-components';
import { action } from '@storybook/addon-actions';
import { select, number, boolean } from '@storybook/addon-knobs';
import { readableColor, toColorString } from 'polished';

import colors from '../../constants/colors';
import RangeSlider, { SlideRail } from './RangeSlider';
import Card, { NoPaddingHeader } from '../Card/Card';

const Row = styled.div`
  display: flex;
  font-family: Gotham, Roboto, sans-serif;
  flex-flow: row nowrap;
  align-items: center;
  padding: .5rem;
  min-width: 20rem;
  width: 90%;
  margin: 0 auto;
`;

export default {
  title: 'RangeSlider',
  component: RangeSlider
};

/* Default */

const Default = () => {
  const [val, setVal] = useState(0);

  const storyValue = number('values', val, {
    range: true,
    min: 0,
    max: 5,
    step: 1,
  });

  useEffect(() => {
    setVal(storyValue);
  }, [storyValue]);

  return (
    <Row>
      <RangeSlider
        disabled={boolean('disabled', false)}
        showDomainLabels={boolean('showDomainLabels', false)}
        showSelectedRange={boolean('showSelectedRange', true)}
        motionBlur={boolean('motionBlur', false)}
        springOnRelease={boolean('springOnRelease', true)}
        min={number('min', 0, {
          range: true,
          min: -10,
          max: 10,
          step: 1,
        })}
        max={number('max', 5, {
          range: true,
          min: -10,
          max: 10,
          step: 1,
        })}
        debounceInterval={number('debounceInterval', 8, {
          range: true,
          min: 0,
          max: 100,
          step: 1,
        })}
        onDrag={(val: number) => {
          setVal(Math.round(val));
          action('onDrag')(val);
        }}
        axisLock={select('axisLock', ['x', 'y', null], 'x')}
        values={[
          {
            value: val,
            label: val
          }
        ]}
      />
    </ Row>
  );
};

Default.design = {
  type: 'figma',
  url: 'https://www.figma.com/file/3r2G00brulOwr9j7F6JF59/Generic-UI-Style?node-id=126%3A2'
};


/* Rating */

const skillLabels = ['WhatJS?', 'I\'ve used it...', 'We\'re aight', 'I love it', 'I dream in React', 'I am React'];
const skillColors = ['red', 'orangered', 'orange', 'goldenrod', 'yellowgreen', 'forestgreen'];

const StyledSlideRail = styled(SlideRail)`
  filter: grayscale(0.5) brightness(1.3);
  border: .5px solid ${colors.grayLight};
  height: .5rem;
  border-radius: .25rem;
  background-image: linear-gradient(to right, ${skillColors.join(', ')});
`;

const Rating = () => {
  const [val, setVal] = useState(0);

  const storyValue = number('values', val, {
    range: true,
    min: 0,
    max: 5,
    step: 1,
  });

  useEffect(() => {
    setVal(storyValue);
  }, [storyValue]);

  return (
    <Row>
      <span>ReactJS:&nbsp;&nbsp;&nbsp;&nbsp;</span>
      <RangeSlider
        StyledSlideRail={StyledSlideRail}
        disabled={boolean('disabled', false)}
        showDomainLabels={boolean('showDomainLabels', false)}
        showSelectedRange={boolean('showSelectedRange', false)}
        motionBlur={boolean('motionBlur', false)}
        springOnRelease={boolean('springOnRelease', true)}
        min={number('min', 0, {
          range: true,
          min: -10,
          max: 10,
          step: 1,
        })}
        max={number('max', 5, {
          range: true,
          min: -10,
          max: 10,
          step: 1,
        })}
        debounceInterval={number('debounceInterval', 8, {
          range: true,
          min: 0,
          max: 100,
          step: 1,
        })}
        onDrag={(val: number) => {
          setVal(Math.round(val));
          action('onDrag')(val);
        }}
        axisLock={select('axisLock', ['x', 'y', null], 'x')}
        values={[
          {
            value: val,
            label: skillLabels[Math.round(val)],
            color: skillColors[Math.round(val)]
          }
        ]}
      />
    </ Row>
  );
};

Rating.design = {
  type: 'figma',
  url: 'https://www.figma.com/file/3r2G00brulOwr9j7F6JF59/Generic-UI-Style?node-id=126%3A2'
};

/* Color Picker */

const ColorPreview = styled.div`
  height: 12rem;
  width: 34rem;
  transition: color .5s;
  font-size: 5rem;
  display: flex;
  justify-content: center;
  align-items: center;
  text-transform: uppercase;
  font-family: Gotham, Roboto, sans-serif;
  font-weight: bold;
`;

const ColorPicker = () => {
  const [hue, setHue] = useState(0);
  const [sat, setSat] = useState(50);
  const [light, setLight] = useState(50);

  const storyHue = number('hue', hue, {
    range: true,
    min: 0,
    max: 255,
    step: 1,
  });

  const storySat = number('saturation', sat, {
    range: true,
    min: 0,
    max: 99,
    step: 1,
  });

  const storyLight = number('lightness', light, {
    range: true,
    min: 0,
    max: 99,
    step: 1,
  });

  useEffect(() => {
    setHue(storyHue);
  }, [storyHue]);

  useEffect(() => {
    setSat(storySat);
  }, [storySat]);

  useEffect(() => {
    setLight(storyLight);
  }, [storyLight]);

  const allHues   = Array.from({length: 360}, (_, i) => i).map(num => `hsl(${num}, ${sat}%, ${light}%)`);
  const allSats   = [`hsl(${hue}, 0%, ${light}%)`, `hsl(${hue}, 100%, ${light}%`];
  const allLights = [`hsl(${hue}, ${sat}%, 10%)`, `hsl(${hue}, ${sat}%, 90%)`];

  return (
    <Card
      elevation={2}
      StyledHeader={NoPaddingHeader}
      header={(<ColorPreview
        style={{
          backgroundColor: `hsl(${hue},${sat}%,${light}%)`,
          color: readableColor(`hsl(${hue},${sat}%,${light}%)`)
        }}>
        {toColorString({
          hue: hue,
          saturation: sat/100,
          lightness: light/100
        })}
      </ColorPreview>)}
    >
      <Row>
        <span>H:&nbsp;&nbsp;&nbsp;</span>
        <RangeSlider
          StyledSlideRail={forwardRef((props, ref) => <SlideRail ref={ref} {...props} style={{
            backgroundImage: `linear-gradient(to right, ${allHues.join(', ')})`
          }} />)}
          disabled={boolean('disabled', false)}
          showDomainLabels={boolean('showDomainLabels', false)}
          showSelectedRange={false}
          motionBlur
          min={0}
          max={360}
          onDrag={(val: number) => {
            setHue(Math.round(val));
            action('onDrag hue')(val);
          }}
          values={[
            {
              value: hue,
              label: hue,
              color: colors.grayLight
            }
          ]}
        />
      </ Row>
      <Row>
        <span>S:&nbsp;&nbsp;&nbsp;</span>
        <RangeSlider
          StyledSlideRail={forwardRef((props, ref) => <SlideRail ref={ref} {...props} style={{
            backgroundImage: `linear-gradient(to right, ${allSats.join(', ')})`
          }} />)}
          min={0}
          max={100}
          onDrag={(val: number) => {
            setSat(Math.round(val));
            action('onDrag saturation')(val);
          }}
          showDomainLabels={false}
          showSelectedRange={false}
          values={[
            {
              value: sat,
              label: sat,
              color: colors.grayLight
            }
          ]}
        />
      </ Row>
      <Row>
      <span>L:&nbsp;&nbsp;&nbsp;</span>
      <RangeSlider
        StyledSlideRail={forwardRef((props, ref) => <SlideRail ref={ref} {...props} style={{
          backgroundImage: `linear-gradient(to right, ${allLights.join(', ')})`
        }} />)}
        min={0}
        max={100}
        onDrag={(val: number) => {
          setLight(Math.round(val));
          action('onDrag light')(val);
        }}
        showDomainLabels={false}
        showSelectedRange={false}
        values={[
          {
            value: light,
            label: light,
            color: colors.grayLight
          }
        ]}
      />
    </ Row>
  </Card>
  );
};

export { Default, Rating, ColorPicker };