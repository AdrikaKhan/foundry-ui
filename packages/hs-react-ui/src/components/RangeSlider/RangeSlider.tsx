import React, { useMemo } from 'react';
import styled, { StyledComponentBase } from 'styled-components';

import { useSpring, a } from 'react-spring';
import { useDrag } from 'react-use-gesture';
import useMeasure from 'react-use-measure';
import polyfill from '@juggle/resize-observer';

import colors from '../../constants/colors';
import { clamp } from '../../utils/math';

type valueProp = {
  value: number,
  label?: String | number | Node,
  color?: String
};

type containerPorps = { showDomainLabels?: boolean, hasHandleLabels?: boolean, disabled: boolean };
export const Container = styled.div`
  ${({ showDomainLabels, hasHandleLabels, disabled }: containerPorps) => `
    position: relative;
    height: 1rem;
    width: 100%;

    user-select: none;

    transition: filter .1s;

    ${disabled ? `
      filter: grayscale(1) contrast(.5) brightness(1.2);
      pointer-events: none;
    ` : ''}

    ${showDomainLabels ? `
        top: -.25rem;
        margin-top: .5rem;
      ` : ''};

      ${hasHandleLabels ? `
        top: -.5rem;
        margin-top: 1rem;
      ` : ''};
  `}
`;

type handleProps = {
  beingDragged?: boolean,
  velocity?: number,
  color: String
};
export const DragHandle = styled(a.div)`
  ${({
    beingDragged = false,
    color = colors.primary
  }: handleProps) => `
    position: absolute;
    bottom: -.125rem;
    left: -.5rem;

    width: 1rem;
    height: 1rem;

    background-color: ${color};
    color: ${color};
    border: .125rem solid ${colors.background};
    border-radius: 50%;

    cursor: grab;
  `}
`;

type handleLabelProps = {velocity?: number};
export const HandleLabel = styled.div`
  ${({ velocity = 0 }: handleLabelProps) => `
      position: absolute;
      bottom: 100%;
      left: 50%;
      transform: translateX(-50%) rotate(${clamp(velocity, -45, 45)}deg);

      font-weight: bold;
      white-space: nowrap;

      pointer-events: none;
  `}
`;

export const SlideRail = styled.div`
  position: absolute;
  top: 50%;
  transform: translateY(-50%);

  width: 100%;
  height: .5rem;

  overflow: hidden;

  border-radius: .25rem;
  background-color: ${colors.grayXlight};
`;

type selectedRangeProps = { min: number, max: number, values: object[], selectedRange: number[] };
export const SelectedRangeRail = styled.div`
  ${({ min, max, values, selectedRange }: selectedRangeProps) => `
    position: absolute;
    top: 0%;
    height: 100%;
    left: ${((selectedRange[0] - min) / (max - min)) * 100}%;
    right: ${((max - selectedRange[1]) / (max - min)) * 100}%;

    background-color: ${colors.primary};
  `}
`;

type domainLabelProps = { position: "left" | "right" };
export const DomainLabel = styled.div`
  ${({ position }: domainLabelProps) => `
    position: absolute;
    bottom: 100%;
    ${position}: 0rem;
    color: ${colors.grayMedium};
    font-size: .5rem;
  `}
`;

export type RangeSliderProps = {
  StyledContainer?: String & StyledComponentBase<any, {}>,
  StyledDragHandle?: String & StyledComponentBase<any, {}>,
  StyledHandleLabel?: String & StyledComponentBase<any, {}>,
  StyledSlideRail?: String & StyledComponentBase<any, {}>,
  StyledSelectedRangeRail?: String & StyledComponentBase<any, {}>,
  StyledDomainLabel?: String & StyledComponentBase<any, {}>,

  showDomainLabels?: boolean,
  showSelectedRange?: boolean,

  onDrag?: Function,
  disabled?: boolean
  min: number,
  max: number,
  values?: number[] | valueProp[],
  markers?: number[] | valueProp[],
};

export default ({
  StyledContainer = Container,
  StyledDragHandle = DragHandle,
  StyledHandleLabel = HandleLabel,
  StyledSlideRail = SlideRail,
  StyledSelectedRangeRail = SelectedRangeRail,
  StyledDomainLabel = DomainLabel,

  showDomainLabels = true,
  showSelectedRange = true,

  onDrag = (newVal: number) => {console.log(newVal)},
  disabled = false,
  min,
  max,
  values,
}: RangeSliderProps) => {
  let hasHandleLabels = false;
  const processedValues = values.map((val: number | valueProp) => {
    if (typeof val === 'number') {
      return { value: val, label: null };
    } else {
      if (val.hasOwnProperty('label')) {
        hasHandleLabels = true;
      }
      return val;
    }
  });
  const selectedRange = [
    Math.min(...processedValues.map((val: valueProp) => val.value)),
    Math.max(...processedValues.map((val: valueProp) => val.value)),
  ];

  const domain = max - min;

  // get the bounding box of the slider
  const [ref, sliderBounds] = useMeasure({ polyfill });
  const pixelPositions = processedValues.map((val: valueProp) => (val.value / domain) * sliderBounds.width);
  // get the x offset and an animation setter function
  const [{ x, y }, set] = useSpring(() => ({ x: pixelPositions[0], y: 0, config: {friction: 13, tension: 100} }), [values]);
  const bind = useDrag(({ down, movement: [deltaX, deltaY] }) => {
    const delta = (deltaX / sliderBounds.width) * domain;
    const newValue = clamp(delta, min, max);
    onDrag(newValue);

    set({ x: down ? deltaX : pixelPositions[0], y: down ? deltaY : 0, immediate: down, config: {friction: 13, tension: 100}});
  },
  {
    axis: 'x',
    initial: [pixelPositions[0], 0],
    bounds: { left: 0, right: sliderBounds.width + 4, top: -8, bottom: (sliderBounds.height / 2) + 8 },
    rubberband: .1
  });

  return (
    <StyledContainer
      disabled={disabled}
      hasHandleLabels={hasHandleLabels}
      showDomainLabels={showDomainLabels}
    >

      <StyledSlideRail
        ref={ref}
      >
        {showSelectedRange &&
          <StyledSelectedRangeRail
            min={min}
            max={max}
            values={processedValues}
            selectedRange={selectedRange}
          />
        }
      </StyledSlideRail>

      {showDomainLabels &&
        <>
          <StyledDomainLabel position="left">{min}</StyledDomainLabel>
          <StyledDomainLabel position="right">{max}</StyledDomainLabel>
        </>
      }

      {processedValues.map(({ value, label, color }: valueProp, i: number) => (
        <StyledDragHandle
          {...bind()}
          style={{ x, y }}
          color={color}
          key={`handle${i}`}
        >
          <StyledHandleLabel value={value}>{label}</StyledHandleLabel>
        </StyledDragHandle>
      ))}

    </StyledContainer>
  )
};
