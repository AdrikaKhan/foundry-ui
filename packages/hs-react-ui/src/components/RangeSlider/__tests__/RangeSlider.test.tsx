import React from 'react';
import { render, fireEvent, waitFor, configure } from '@testing-library/react';
import Icon from '@mdi/react';
import RangeSlider from '../RangeSlider';

configure({ testIdAttribute: 'data-test-id' });

const testId = 'hs-ui-range-slider-unit-test';

describe('RangeSlider', () => {
  it('renders', async () => {
    const { container, getByTestId } = render(<RangeSlider min={0} max={10} testId="unit-test" />);

    await waitFor(() => getByTestId(testId));

    expect(container).toMatchSnapshot();
  });
});
