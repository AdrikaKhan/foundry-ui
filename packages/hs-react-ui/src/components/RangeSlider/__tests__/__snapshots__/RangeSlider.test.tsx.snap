// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider renders 1`] = `
.c0 {
  position: relative;
  height: 1rem;
  width: 100%;
  font-family: Montserrat,Roboto,sans-serif;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: filter .1s;
  transition: filter .1s;
  top: -.5rem;
  margin-top: 1rem;
}

.c1 {
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  width: 100%;
  height: 0.25rem;
  overflow: hidden;
  border-radius: 0.125rem;
  background-color: #d3d6d7;
}

.c2 {
  position: absolute;
  top: 0%;
  height: 100%;
  left: Infinity%;
  right: Infinity%;
  -webkit-transition: left .3s,right .3s;
  transition: left .3s,right .3s;
  background-color: #e75b27;
}

.c3 {
  position: absolute;
  bottom: 100%;
  left: 0rem;
  color: #51575d;
  font-size: .5rem;
}

.c4 {
  position: absolute;
  bottom: 100%;
  right: 0rem;
  color: #51575d;
  font-size: .5rem;
}

<div>
  <div
    class="c0"
    data-test-id="hs-ui-range-slider-unit-test"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        max="10"
        min="0"
        values=""
      />
    </div>
    <div
      class="c3"
    >
      0
    </div>
    <div
      class="c4"
    >
      10
    </div>
  </div>
</div>
`;
